### PR TITLE
Fix coloring and hover data for nonuniform heatmap & contour bricks

### LINKED ISSUE #2233: Invalid heatmap values
This heatmap shows that my values are (15, 22.75), (22.75, 28), (26, 11), the first two are wrong.

https://jsfiddle.net/zc5r6v6t/3/

plotly

### PR DESCRIPTION
**Pull request by alexcjohnson**
Fixes #2233 - we were finding brick edges (as midpoints between data values) and then forever after using the midpoints of bricks as the hover data and (heatmap with `zsmooth: 'best'` or contour with `coloring: 'heatmap'`) the point of exact color matching, rather than using the original data values for these purposes.

@dfcreative one more bugfix PR to review before I move onto building new stuff?

### CODE REVIEW

File: src/traces/heatmap/plot.js [hunk 1]
```diff
@@ -415,3 +383,64 @@ function plotOne(gd, plotinfo, cd) {
...
+        bin0: bin0,
+        frac: frac,
+        bin1: Math.round(bin0 + frac / (interp - bin0))
+    };
+}
+
+function findInterpFromCenters(pixel, centerPixArray) {
+    // if(pixel <= centerPixArray[0]) return {bin0: 0, bin1: 0, frac: 0};
+    var maxBin = centerPixArray.length - 1;
+    // if(pixel >= centerPixArray[lastCenter]) return {bin0: lastCenter, bin1: lastCenter, frac: 0};
```
Comment (dfcreative) [inline src/traces/heatmap/plot.js]: Are these being temporarily commented out?

**Comment by dfcreative** [comment]
@alexcjohnson is that ok that we see `NaN` in empty blocks?

image

File: src/traces/heatmap/plot.js (hunk 1 above)
Comment (alexcjohnson) [reply src/traces/heatmap/plot.js]: Ah thanks - those are obsolete -> https://github.com/plotly/plotly.js/pull/2288/commits/3581c87fbe9b151162dc854992a963aa3599702d

**Comment by alexcjohnson** [comment]
> is that ok that we see `NaN` in empty blocks?

I think so - it's certainly better than saying `0`. Might be nice to be able to customize what NaN displays as, but this is a reasonable default I think. See also #975

**Comment by alexcjohnson** [comment]
> Not relevant to this PR, but is that fine that hover labels appear at different positions than with `contour`

I believe this behaves as expected. For a heatmap, the hover label should not point to the data value but to the edge of the brick, whereas contour map labels should point exactly to the data value.

Heatmaps you're defining, either implicitly or explicitly, bricks of a certain value, and you're saying "this is the value that applies to this brick". That metaphor gets a little bit muddied by smoothing the heatmap, but that's the idea anyway. Whereas with contour maps, you're implying that there's some continuous function and you've only sampled it at certain locations, the data points, but it has a distinct value at every (x,y) pair. But we don't extrapolate beyond the range of the data, even as far as the edge of the corresponding heatmap brick.

**Comment by dfcreative** [comment]
Looks good then! :dancer:

**Pull request merged**